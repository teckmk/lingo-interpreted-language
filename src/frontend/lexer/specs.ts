import { EOL } from "os"

export enum TokenType {
  NumberLiteral = "NUMBER_LITERAL",
  StringLiteral = "STRING_LITERAL",
  Identifier = "IDENTIFIER",
  SingleLineComment = "SINLE_LINE_COMMENT",
  DocComment = "DOC_COMMENT",
  MultiLineComment = "MULTI_LINE_COMMENT",
  WhiteSpace = "WHITE_SPACE",
  Tab = "TAB",
  Indent = "INDENT",
  Dedent = "DEDENT",
  EOF = "END_OF_FILE",
  EOL = "END_OF_LINE",

  Equals = "ASSIGNMENT_OPERATOR", // =
  Comma = "COMMA", // ,
  Dot = "DOT", // .
  Exclamation = "EXLAMATION", // "!"
  Colon = "COLON", //":"
  OpenParen = "OPEN_PAREN", //  (
  CloseParen = "CLOSE_PAREN", // )
  OpenBrace = "OPEN_BRACE", // {
  CloseBrace = "CLOSE_BRACE", // }
  OpenBracket = "OPEN_BRACKET", // [
  CloseBracket = "CLOSE_BRACKET", // ]
  LogicGate = "LOGIC_GATE", // && || and or
  AdditiveOperator = "ADDITIVE_OPERATOR", //  + -
  MulitipicativeOperator = "MULTIPICATIVE_OPERATOR", // / * %
  EqualityOperator = "EQUALITY_OPERATOR", // == !=
  RelationalOperator = "RELATIONAL_OPERATOR", // > < >= <=

  // Reserved
  // KeyWords
  Let = "LET",
  Const = "CONST",
  Final = "FINAL",
  Fn = "FN",
  Return = "RETURN",
  If = "IF",
  Else = "ELSE",
  While = "WHILE",

  // Primitive Types
  NumberType = "NUMBER_TYPE",
  StringType = "STRING_TYPE",
  BooleanType = "BOOL_TYPE",
  DynamicType = "DYNAMIC_TYPE",
}

export type Spec = {
  regex: RegExp
  tokenType: TokenType
}

export const specs: Spec[] = [
  { regex: /^[+-]?([\d]*[.])?[\d]+/, tokenType: TokenType.NumberLiteral },
  { regex: /^"[^"]*"/, tokenType: TokenType.StringLiteral },
  { regex: /^'[^']*'/, tokenType: TokenType.StringLiteral },

  { regex: /^\bvar\b/, tokenType: TokenType.Let },
  { regex: /^\bconst\b/, tokenType: TokenType.Const },
  { regex: /^\bfinal\b/, tokenType: TokenType.Final },
  { regex: /^\bfn\b/, tokenType: TokenType.Fn },
  { regex: /^\breturn\b/, tokenType: TokenType.Return },
  { regex: /^\bif\b/, tokenType: TokenType.If },
  { regex: /^\belse\b/, tokenType: TokenType.Else },
  { regex: /^\bwhile\b/, tokenType: TokenType.While },

  { regex: /^\bnumber\b/, tokenType: TokenType.NumberType },
  { regex: /^\bstring\b/, tokenType: TokenType.StringType },
  { regex: /^\bbool\b/, tokenType: TokenType.BooleanType },
  { regex: /^\bdynamic\b/, tokenType: TokenType.DynamicType },

  { regex: /^[a-zA-Z_][a-zA-Z0-9_]*/, tokenType: TokenType.Identifier },

  { regex: /^\/\/.*/, tokenType: TokenType.SingleLineComment },
  { regex: /^\/\/\/.*/, tokenType: TokenType.DocComment },

  // these are making problem with line numbers, because they are read as single line, no matter how many lines they span on....
  // { regex: /^\/\*[\s\S]*?\*\//, tokenType: TokenType.MultiLineComment },

  // operators
  { regex: /^[+-]/, tokenType: TokenType.AdditiveOperator },
  { regex: /^[*/%]/, tokenType: TokenType.MulitipicativeOperator },
  { regex: /^[><]=?/, tokenType: TokenType.RelationalOperator },
  { regex: /^[=!]=/, tokenType: TokenType.EqualityOperator },

  { regex: /^&&/, tokenType: TokenType.LogicGate },
  { regex: /^\|\|/, tokenType: TokenType.LogicGate },
  { regex: /^\band\b/, tokenType: TokenType.LogicGate },
  { regex: /^\bor\b/, tokenType: TokenType.LogicGate },

  { regex: /^:/, tokenType: TokenType.Colon },
  { regex: /^,/, tokenType: TokenType.Comma },
  { regex: /^\./, tokenType: TokenType.Dot },
  { regex: /^=/, tokenType: TokenType.Equals },
  { regex: /^!/, tokenType: TokenType.Exclamation },
  { regex: /^\(/, tokenType: TokenType.OpenParen },
  { regex: /^\)/, tokenType: TokenType.CloseParen },
  { regex: /^{/, tokenType: TokenType.OpenBrace },
  { regex: /^}/, tokenType: TokenType.CloseBrace },
  { regex: /^\[/, tokenType: TokenType.OpenBracket },
  { regex: /^\]/, tokenType: TokenType.CloseBracket },

  { regex: RegExp(`^${EOL}`), tokenType: TokenType.EOL },

  { regex: /^[^\S\r\n]+/, tokenType: TokenType.WhiteSpace }, // match one or more spaces and tabs
]
