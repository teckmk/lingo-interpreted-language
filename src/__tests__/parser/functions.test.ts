import { specs } from "../../frontend/lexer/specs"
import { tokenize } from "../../frontend/lexer/tokenizer"
import Parser from "../../frontend/parser"

describe("FunctionDeclaration", () => {
  it("should parse function declarations without parameters", () => {
    const code = "fn a() {}"
    const tokens = tokenize(specs, "test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "FunctionDeclaration",
          name: "a",
          parameters: [],
          body: [],
        },
      ],
    })
  })

  it("should parse function declarations with parameters", () => {
    const code = "fn a(b: number, c: string) {}"
    const tokens = tokenize(specs, "test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "FunctionDeclaration",
          name: "a",
          parameters: [
            {
              kind: "FunctionParam",
              name: "b",
              type: "number",
            },
            {
              kind: "FunctionParam",
              name: "c",
              type: "string",
            },
          ],
          body: [],
        },
      ],
    })
  })

  it("should parse function declarations with return type", () => {
    const code = "fn a() -> number {}"
    const tokens = tokenize(specs, "test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "FunctionDeclaration",
          name: "a",
          parameters: [],
          returnType: "number",
          body: [],
        },
      ],
    })
  })

  it("should parse function declarations with parameters and return type", () => {
    const code = "fn a(b: number, c: string) -> number {}"
    const tokens = tokenize(specs, "test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "FunctionDeclaration",
          name: "a",
          parameters: [
            {
              kind: "FunctionParam",
              name: "b",
              type: "number",
            },
            {
              kind: "FunctionParam",
              name: "c",
              type: "string",
            },
          ],
          returnType: "number",
          body: [],
        },
      ],
    })
  })

  it("should parse function declarations with body", () => {
    const code = "fn a() { return 1 }"
    const tokens = tokenize(specs, "test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "FunctionDeclaration",
          name: "a",
          parameters: [],
          body: [
            {
              kind: "ReturnStatement",
              value: {
                kind: "NumericLiteral",
                value: 1,
              },
            },
          ],
        },
      ],
    })
  })

  it("should parse function declarations with parameters and body", () => {
    const code = "fn a(b: number, c: string) { return 1 }"
    const tokens = tokenize(specs, "test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "FunctionDeclaration",
          name: "a",
          parameters: [
            {
              kind: "FunctionParam",
              name: "b",
              type: "number",
            },
            {
              kind: "FunctionParam",
              name: "c",
              type: "string",
            },
          ],
          body: [
            {
              kind: "ReturnStatement",
              value: {
                kind: "NumericLiteral",
                value: 1,
              },
            },
          ],
        },
      ],
    })
  })

  it("should parse function declarations with return type and body", () => {
    const code = "fn a() -> number { return 1 }"
    const tokens = tokenize(specs, "test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "FunctionDeclaration",
          name: "a",
          parameters: [],
          returnType: "number",
          body: [
            {
              kind: "ReturnStatement",
              value: {
                kind: "NumericLiteral",
                value: 1,
              },
            },
          ],
        },
      ],
    })
  })

  it("should parse function declarations with parameters, return type and body", () => {
    const code = "fn a(b: number, c: string) -> number { return 1 }"
    const tokens = tokenize(specs, "test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "FunctionDeclaration",
          name: "a",
          parameters: [
            {
              kind: "FunctionParam",
              name: "b",
              type: "number",
            },
            {
              kind: "FunctionParam",
              name: "c",
              type: "string",
            },
          ],
          returnType: "number",
          body: [
            {
              kind: "ReturnStatement",
              value: {
                kind: "NumericLiteral",
                value: 1,
              },
            },
          ],
        },
      ],
    })
  })

  it("should parse function declarations with multiple statements in body", () => {
    const code = `fn a() { 
      var b = 1
      return b 
    }`
    const tokens = tokenize(specs, "test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "FunctionDeclaration",
          name: "a",
          parameters: [],
          body: [
            {
              kind: "VarDeclaration",
              identifier: "b",
              modifier: "variable",
              type: undefined, // type is not explicitly defined
              value: {
                kind: "NumericLiteral",
                value: 1,
              },
            },
            {
              kind: "ReturnStatement",
              value: {
                kind: "Identifier",
                symbol: "b",
              },
            },
          ],
        },
      ],
    })
  })

  it("should parse function declarations with indented code block", () => {
    const code = `fn a(): 
      var b = 1
      return b`
    const tokens = tokenize(specs, "test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "FunctionDeclaration",
          name: "a",
          parameters: [],
          body: [
            {
              kind: "VarDeclaration",
              identifier: "b",
              modifier: "variable",
              type: undefined, // type is not explicitly defined
              value: {
                kind: "NumericLiteral",
                value: 1,
              },
            },
            {
              kind: "ReturnStatement",
              value: {
                kind: "Identifier",
                symbol: "b",
              },
            },
          ],
        },
      ],
    })
  })

  it("should parse function declarations with indented code block and return type", () => {
    const code = `fn a() -> number: 
      var b = 1
      return b`
    const tokens = tokenize(specs, "test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "FunctionDeclaration",
          name: "a",
          parameters: [],
          returnType: "number",
          body: [
            {
              kind: "VarDeclaration",
              identifier: "b",
              modifier: "variable",
              type: undefined, // type is not explicitly defined
              value: {
                kind: "NumericLiteral",
                value: 1,
              },
            },
            {
              kind: "ReturnStatement",
              value: {
                kind: "Identifier",
                symbol: "b",
              },
            },
          ],
        },
      ],
    })
  })
})
