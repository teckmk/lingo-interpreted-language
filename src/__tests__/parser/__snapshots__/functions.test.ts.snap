// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser - FunctionDeclaration should parse function declarations with body 1`] = `
{
  "body": [
    {
      "body": [
        {
          "kind": "ReturnStatement",
          "value": {
            "kind": "NumericLiteral",
            "value": {
              "position": {
                "end": {
                  "column": 18,
                  "line": 1,
                },
                "start": {
                  "column": 17,
                  "line": 1,
                },
              },
              "value": 1,
            },
          },
        },
      ],
      "kind": "FunctionDeclaration",
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": false,
          "isMethod": false,
          "isStatic": false,
        },
        "name": {
          "position": {
            "end": {
              "column": 5,
              "line": 1,
            },
            "start": {
              "column": 4,
              "line": 1,
            },
          },
          "value": "a",
        },
        "parameters": [],
        "returnType": undefined,
        "typeParameters": undefined,
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - FunctionDeclaration should parse function declarations with indented code block 1`] = `
{
  "body": [
    {
      "body": [
        {
          "kind": "Identifier",
          "symbol": {
            "position": {
              "end": {
                "column": 10,
                "line": 2,
              },
              "start": {
                "column": 7,
                "line": 2,
              },
            },
            "value": "var",
          },
        },
        {
          "assigne": {
            "kind": "Identifier",
            "symbol": {
              "position": {
                "end": {
                  "column": 12,
                  "line": 2,
                },
                "start": {
                  "column": 11,
                  "line": 2,
                },
              },
              "value": "b",
            },
          },
          "kind": "AssignmentExpr",
          "value": {
            "kind": "NumericLiteral",
            "value": {
              "position": {
                "end": {
                  "column": 16,
                  "line": 2,
                },
                "start": {
                  "column": 15,
                  "line": 2,
                },
              },
              "value": 1,
            },
          },
        },
        {
          "kind": "ReturnStatement",
          "value": {
            "kind": "Identifier",
            "symbol": {
              "position": {
                "end": {
                  "column": 15,
                  "line": 3,
                },
                "start": {
                  "column": 14,
                  "line": 3,
                },
              },
              "value": "b",
            },
          },
        },
      ],
      "kind": "FunctionDeclaration",
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": false,
          "isMethod": false,
          "isStatic": false,
        },
        "name": {
          "position": {
            "end": {
              "column": 5,
              "line": 1,
            },
            "start": {
              "column": 4,
              "line": 1,
            },
          },
          "value": "a",
        },
        "parameters": [],
        "returnType": undefined,
        "typeParameters": undefined,
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - FunctionDeclaration should parse function declarations with indented code block and return type 1`] = `
{
  "body": [
    {
      "body": [
        {
          "kind": "Identifier",
          "symbol": {
            "position": {
              "end": {
                "column": 10,
                "line": 2,
              },
              "start": {
                "column": 7,
                "line": 2,
              },
            },
            "value": "var",
          },
        },
        {
          "assigne": {
            "kind": "Identifier",
            "symbol": {
              "position": {
                "end": {
                  "column": 12,
                  "line": 2,
                },
                "start": {
                  "column": 11,
                  "line": 2,
                },
              },
              "value": "b",
            },
          },
          "kind": "AssignmentExpr",
          "value": {
            "kind": "NumericLiteral",
            "value": {
              "position": {
                "end": {
                  "column": 16,
                  "line": 2,
                },
                "start": {
                  "column": 15,
                  "line": 2,
                },
              },
              "value": 1,
            },
          },
        },
        {
          "kind": "ReturnStatement",
          "value": {
            "kind": "Identifier",
            "symbol": {
              "position": {
                "end": {
                  "column": 15,
                  "line": 3,
                },
                "start": {
                  "column": 14,
                  "line": 3,
                },
              },
              "value": "b",
            },
          },
        },
      ],
      "kind": "FunctionDeclaration",
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": false,
          "isMethod": false,
          "isStatic": false,
        },
        "name": {
          "position": {
            "end": {
              "column": 5,
              "line": 1,
            },
            "start": {
              "column": 4,
              "line": 1,
            },
          },
          "value": "a",
        },
        "parameters": [],
        "returnType": {
          "isNominal": false,
          "kind": "PrimitiveType",
          "name": {
            "position": {
              "end": {
                "column": 17,
                "line": 1,
              },
              "start": {
                "column": 11,
                "line": 1,
              },
            },
            "value": "number",
          },
        },
        "typeParameters": undefined,
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - FunctionDeclaration should parse function declarations with multiple return types 1`] = `
{
  "body": [
    {
      "body": [],
      "kind": "FunctionDeclaration",
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": false,
          "isMethod": false,
          "isStatic": false,
        },
        "name": {
          "position": {
            "end": {
              "column": 5,
              "line": 1,
            },
            "start": {
              "column": 4,
              "line": 1,
            },
          },
          "value": "a",
        },
        "parameters": [],
        "returnType": [
          {
            "isNominal": false,
            "kind": "PrimitiveType",
            "name": {
              "position": {
                "end": {
                  "column": 17,
                  "line": 1,
                },
                "start": {
                  "column": 11,
                  "line": 1,
                },
              },
              "value": "number",
            },
          },
          {
            "isNominal": true,
            "kind": "PrimitiveType",
            "name": {
              "position": {
                "end": {
                  "column": 25,
                  "line": 1,
                },
                "start": {
                  "column": 19,
                  "line": 1,
                },
              },
              "value": "string",
            },
          },
        ],
        "typeParameters": undefined,
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - FunctionDeclaration should parse function declarations with multiple return values 1`] = `
{
  "body": [
    {
      "body": [
        {
          "kind": "ReturnStatement",
          "value": [
            {
              "kind": "NumericLiteral",
              "value": {
                "position": {
                  "end": {
                    "column": 15,
                    "line": 2,
                  },
                  "start": {
                    "column": 14,
                    "line": 2,
                  },
                },
                "value": 1,
              },
            },
            {
              "kind": "NumericLiteral",
              "value": {
                "position": {
                  "end": {
                    "column": 18,
                    "line": 2,
                  },
                  "start": {
                    "column": 17,
                    "line": 2,
                  },
                },
                "value": 2,
              },
            },
          ],
        },
      ],
      "kind": "FunctionDeclaration",
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": false,
          "isMethod": false,
          "isStatic": false,
        },
        "name": {
          "position": {
            "end": {
              "column": 5,
              "line": 1,
            },
            "start": {
              "column": 4,
              "line": 1,
            },
          },
          "value": "a",
        },
        "parameters": [],
        "returnType": [
          {
            "isNominal": false,
            "kind": "PrimitiveType",
            "name": {
              "position": {
                "end": {
                  "column": 17,
                  "line": 1,
                },
                "start": {
                  "column": 11,
                  "line": 1,
                },
              },
              "value": "number",
            },
          },
          {
            "isNominal": true,
            "kind": "PrimitiveType",
            "name": {
              "position": {
                "end": {
                  "column": 25,
                  "line": 1,
                },
                "start": {
                  "column": 19,
                  "line": 1,
                },
              },
              "value": "number",
            },
          },
        ],
        "typeParameters": undefined,
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - FunctionDeclaration should parse function declarations with multiple statements in body 1`] = `
{
  "body": [
    {
      "body": [
        {
          "kind": "Identifier",
          "symbol": {
            "position": {
              "end": {
                "column": 10,
                "line": 2,
              },
              "start": {
                "column": 7,
                "line": 2,
              },
            },
            "value": "var",
          },
        },
        {
          "assigne": {
            "kind": "Identifier",
            "symbol": {
              "position": {
                "end": {
                  "column": 12,
                  "line": 2,
                },
                "start": {
                  "column": 11,
                  "line": 2,
                },
              },
              "value": "b",
            },
          },
          "kind": "AssignmentExpr",
          "value": {
            "kind": "NumericLiteral",
            "value": {
              "position": {
                "end": {
                  "column": 16,
                  "line": 2,
                },
                "start": {
                  "column": 15,
                  "line": 2,
                },
              },
              "value": 1,
            },
          },
        },
        {
          "kind": "ReturnStatement",
          "value": {
            "kind": "Identifier",
            "symbol": {
              "position": {
                "end": {
                  "column": 15,
                  "line": 3,
                },
                "start": {
                  "column": 14,
                  "line": 3,
                },
              },
              "value": "b",
            },
          },
        },
      ],
      "kind": "FunctionDeclaration",
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": false,
          "isMethod": false,
          "isStatic": false,
        },
        "name": {
          "position": {
            "end": {
              "column": 5,
              "line": 1,
            },
            "start": {
              "column": 4,
              "line": 1,
            },
          },
          "value": "a",
        },
        "parameters": [],
        "returnType": undefined,
        "typeParameters": undefined,
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - FunctionDeclaration should parse function declarations with parameters 1`] = `
{
  "body": [
    {
      "body": [],
      "kind": "FunctionDeclaration",
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": false,
          "isMethod": false,
          "isStatic": false,
        },
        "name": {
          "position": {
            "end": {
              "column": 5,
              "line": 1,
            },
            "start": {
              "column": 4,
              "line": 1,
            },
          },
          "value": "a",
        },
        "parameters": [
          {
            "kind": "FunctionParam",
            "name": {
              "position": {
                "end": {
                  "column": 7,
                  "line": 1,
                },
                "start": {
                  "column": 6,
                  "line": 1,
                },
              },
              "value": "b",
            },
            "type": {
              "isNominal": false,
              "kind": "PrimitiveType",
              "name": {
                "position": {
                  "end": {
                    "column": 15,
                    "line": 1,
                  },
                  "start": {
                    "column": 9,
                    "line": 1,
                  },
                },
                "value": "number",
              },
            },
          },
          {
            "kind": "FunctionParam",
            "name": {
              "position": {
                "end": {
                  "column": 18,
                  "line": 1,
                },
                "start": {
                  "column": 17,
                  "line": 1,
                },
              },
              "value": "c",
            },
            "type": {
              "isNominal": false,
              "kind": "PrimitiveType",
              "name": {
                "position": {
                  "end": {
                    "column": 26,
                    "line": 1,
                  },
                  "start": {
                    "column": 20,
                    "line": 1,
                  },
                },
                "value": "string",
              },
            },
          },
        ],
        "returnType": undefined,
        "typeParameters": undefined,
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - FunctionDeclaration should parse function declarations with parameters and body 1`] = `
{
  "body": [
    {
      "body": [
        {
          "kind": "ReturnStatement",
          "value": {
            "kind": "NumericLiteral",
            "value": {
              "position": {
                "end": {
                  "column": 38,
                  "line": 1,
                },
                "start": {
                  "column": 37,
                  "line": 1,
                },
              },
              "value": 1,
            },
          },
        },
      ],
      "kind": "FunctionDeclaration",
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": false,
          "isMethod": false,
          "isStatic": false,
        },
        "name": {
          "position": {
            "end": {
              "column": 5,
              "line": 1,
            },
            "start": {
              "column": 4,
              "line": 1,
            },
          },
          "value": "a",
        },
        "parameters": [
          {
            "kind": "FunctionParam",
            "name": {
              "position": {
                "end": {
                  "column": 7,
                  "line": 1,
                },
                "start": {
                  "column": 6,
                  "line": 1,
                },
              },
              "value": "b",
            },
            "type": {
              "isNominal": false,
              "kind": "PrimitiveType",
              "name": {
                "position": {
                  "end": {
                    "column": 15,
                    "line": 1,
                  },
                  "start": {
                    "column": 9,
                    "line": 1,
                  },
                },
                "value": "number",
              },
            },
          },
          {
            "kind": "FunctionParam",
            "name": {
              "position": {
                "end": {
                  "column": 18,
                  "line": 1,
                },
                "start": {
                  "column": 17,
                  "line": 1,
                },
              },
              "value": "c",
            },
            "type": {
              "isNominal": false,
              "kind": "PrimitiveType",
              "name": {
                "position": {
                  "end": {
                    "column": 26,
                    "line": 1,
                  },
                  "start": {
                    "column": 20,
                    "line": 1,
                  },
                },
                "value": "string",
              },
            },
          },
        ],
        "returnType": undefined,
        "typeParameters": undefined,
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - FunctionDeclaration should parse function declarations with parameters and return type 1`] = `
{
  "body": [
    {
      "body": [],
      "kind": "FunctionDeclaration",
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": false,
          "isMethod": false,
          "isStatic": false,
        },
        "name": {
          "position": {
            "end": {
              "column": 5,
              "line": 1,
            },
            "start": {
              "column": 4,
              "line": 1,
            },
          },
          "value": "a",
        },
        "parameters": [
          {
            "kind": "FunctionParam",
            "name": {
              "position": {
                "end": {
                  "column": 7,
                  "line": 1,
                },
                "start": {
                  "column": 6,
                  "line": 1,
                },
              },
              "value": "b",
            },
            "type": {
              "isNominal": false,
              "kind": "PrimitiveType",
              "name": {
                "position": {
                  "end": {
                    "column": 15,
                    "line": 1,
                  },
                  "start": {
                    "column": 9,
                    "line": 1,
                  },
                },
                "value": "number",
              },
            },
          },
          {
            "kind": "FunctionParam",
            "name": {
              "position": {
                "end": {
                  "column": 18,
                  "line": 1,
                },
                "start": {
                  "column": 17,
                  "line": 1,
                },
              },
              "value": "c",
            },
            "type": {
              "isNominal": false,
              "kind": "PrimitiveType",
              "name": {
                "position": {
                  "end": {
                    "column": 26,
                    "line": 1,
                  },
                  "start": {
                    "column": 20,
                    "line": 1,
                  },
                },
                "value": "string",
              },
            },
          },
        ],
        "returnType": {
          "isNominal": false,
          "kind": "PrimitiveType",
          "name": {
            "position": {
              "end": {
                "column": 37,
                "line": 1,
              },
              "start": {
                "column": 31,
                "line": 1,
              },
            },
            "value": "number",
          },
        },
        "typeParameters": undefined,
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - FunctionDeclaration should parse function declarations with parameters, return type and body 1`] = `
{
  "body": [
    {
      "body": [
        {
          "kind": "ReturnStatement",
          "value": {
            "kind": "NumericLiteral",
            "value": {
              "position": {
                "end": {
                  "column": 48,
                  "line": 1,
                },
                "start": {
                  "column": 47,
                  "line": 1,
                },
              },
              "value": 1,
            },
          },
        },
      ],
      "kind": "FunctionDeclaration",
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": false,
          "isMethod": false,
          "isStatic": false,
        },
        "name": {
          "position": {
            "end": {
              "column": 5,
              "line": 1,
            },
            "start": {
              "column": 4,
              "line": 1,
            },
          },
          "value": "a",
        },
        "parameters": [
          {
            "kind": "FunctionParam",
            "name": {
              "position": {
                "end": {
                  "column": 7,
                  "line": 1,
                },
                "start": {
                  "column": 6,
                  "line": 1,
                },
              },
              "value": "b",
            },
            "type": {
              "isNominal": false,
              "kind": "PrimitiveType",
              "name": {
                "position": {
                  "end": {
                    "column": 15,
                    "line": 1,
                  },
                  "start": {
                    "column": 9,
                    "line": 1,
                  },
                },
                "value": "number",
              },
            },
          },
          {
            "kind": "FunctionParam",
            "name": {
              "position": {
                "end": {
                  "column": 18,
                  "line": 1,
                },
                "start": {
                  "column": 17,
                  "line": 1,
                },
              },
              "value": "c",
            },
            "type": {
              "isNominal": false,
              "kind": "PrimitiveType",
              "name": {
                "position": {
                  "end": {
                    "column": 26,
                    "line": 1,
                  },
                  "start": {
                    "column": 20,
                    "line": 1,
                  },
                },
                "value": "string",
              },
            },
          },
        ],
        "returnType": {
          "isNominal": false,
          "kind": "PrimitiveType",
          "name": {
            "position": {
              "end": {
                "column": 37,
                "line": 1,
              },
              "start": {
                "column": 31,
                "line": 1,
              },
            },
            "value": "number",
          },
        },
        "typeParameters": undefined,
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - FunctionDeclaration should parse function declarations with return type 1`] = `
{
  "body": [
    {
      "body": [],
      "kind": "FunctionDeclaration",
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": false,
          "isMethod": false,
          "isStatic": false,
        },
        "name": {
          "position": {
            "end": {
              "column": 5,
              "line": 1,
            },
            "start": {
              "column": 4,
              "line": 1,
            },
          },
          "value": "a",
        },
        "parameters": [],
        "returnType": {
          "isNominal": false,
          "kind": "PrimitiveType",
          "name": {
            "position": {
              "end": {
                "column": 17,
                "line": 1,
              },
              "start": {
                "column": 11,
                "line": 1,
              },
            },
            "value": "number",
          },
        },
        "typeParameters": undefined,
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - FunctionDeclaration should parse function declarations with return type and body 1`] = `
{
  "body": [
    {
      "body": [
        {
          "kind": "ReturnStatement",
          "value": {
            "kind": "NumericLiteral",
            "value": {
              "position": {
                "end": {
                  "column": 28,
                  "line": 1,
                },
                "start": {
                  "column": 27,
                  "line": 1,
                },
              },
              "value": 1,
            },
          },
        },
      ],
      "kind": "FunctionDeclaration",
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": false,
          "isMethod": false,
          "isStatic": false,
        },
        "name": {
          "position": {
            "end": {
              "column": 5,
              "line": 1,
            },
            "start": {
              "column": 4,
              "line": 1,
            },
          },
          "value": "a",
        },
        "parameters": [],
        "returnType": {
          "isNominal": false,
          "kind": "PrimitiveType",
          "name": {
            "position": {
              "end": {
                "column": 17,
                "line": 1,
              },
              "start": {
                "column": 11,
                "line": 1,
              },
            },
            "value": "number",
          },
        },
        "typeParameters": undefined,
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - FunctionDeclaration should parse function declarations without parameters 1`] = `
{
  "body": [
    {
      "body": [],
      "kind": "FunctionDeclaration",
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": false,
          "isMethod": false,
          "isStatic": false,
        },
        "name": {
          "position": {
            "end": {
              "column": 5,
              "line": 1,
            },
            "start": {
              "column": 4,
              "line": 1,
            },
          },
          "value": "a",
        },
        "parameters": [],
        "returnType": undefined,
        "typeParameters": undefined,
      },
    },
  ],
  "kind": "Program",
}
`;
